### PR TITLE
ADEN-7608 Add Labrador:Basset functionality

### PR DESCRIPTION
**Pull request by dmnsobczak**

### CODE REVIEW

File: src/ad-engine/utils/geo.js [hunk 1]
```diff
@@ -1,54 +1,106 @@
...
 	samplingSeparator = '/';
 
-let geoData = null,
-	cache = {};
+let cache = {},
+	cookieLoaded = false,
+	geoData = null,
+	sessionId = '';
+
+function hasCookie(countryList) {
```
Comment (fraszczakszymon) [inline src/ad-engine/utils/geo.js]: s/hasCookie/hasCache?

File: src/ad-engine/utils/geo.js [hunk 2]
```diff
@@ -1,54 +1,106 @@
 import Cookies from 'js-cookie';
 import Random from './random';
 
-const earth = 'XX',
+const cookieMarker = '-cached',
```
Comment (fraszczakszymon) [inline src/ad-engine/utils/geo.js]: What about `cacheMarker`?

File: src/ad-engine/utils/geo.js [hunk 3]
```diff
@@ -1,54 +1,106 @@
...
+		withCookie
 	};
+
+	if (withCookie) {
+		synchronizeCookie();
+	}
+}
+
+function readSession() {
+	sessionId = sessionId || Cookies.get('wikia_session_id') || 'ae3';
```
Comment (fraszczakszymon) [inline src/ad-engine/utils/geo.js]: Can we get name of cookie from context? e.g. `options.sessionIdCookieName`

File: src/ad-engine/utils/geo.js [hunk 4]
```diff
@@ -1,54 +1,106 @@
...
+
+	Object.keys(cache).forEach((variable) => {
+		if (cache[variable].withCookie) {
+			cachedVariables[variable] = cache[variable];
+		}
+	});
+
+	Cookies.set(`${sessionId}_basset`, JSON.stringify(cachedVariables), {
+		path: '/',
+		domain: domain.length > 1 ? `.${domain[domain.length - 2]}.${domain[domain.length - 1]}` : undefined,
```
Comment (fraszczakszymon) [inline src/ad-engine/utils/geo.js]: What about `context.get('options.session.cookieDomain')`?

File: src/ad-engine/utils/geo.js (hunk 3 above)
Comment (fraszczakszymon) [reply src/ad-engine/utils/geo.js]: What about storing it in the context? `context.get('options.session.id')`

File: src/ad-engine/utils/geo.js [hunk 5]
```diff
@@ -1,54 +1,106 @@
...
+	const cachedVariables = {};
+	const domain = (window.location.hostname).split('.');
+
+	Object.keys(cache).forEach((variable) => {
+		if (cache[variable].withCookie) {
+			cachedVariables[variable] = cache[variable];
+		}
+	});
+
+	Cookies.set(`${sessionId}_basset`, JSON.stringify(cachedVariables), {
```
Comment (fraszczakszymon) [inline src/ad-engine/utils/geo.js]: Correct me if I am wrong, but it means that cookie expires with closed browser, right? Session cookie expires after 30mins. Basset cookie should last longer/the same (configurable) time

File: src/ad-engine/utils/geo.js (hunk 2 above)
Comment (dmnsobczak) [reply src/ad-engine/utils/geo.js]: I used that name at the first attempt :D Renamed.

File: src/ad-engine/utils/geo.js (hunk 1 above)
Comment (dmnsobczak) [reply src/ad-engine/utils/geo.js]: Si, renamed.

File: src/ad-engine/utils/geo.js (hunk 4 above)
Comment (dmnsobczak) [reply src/ad-engine/utils/geo.js]: Discussed offline: it's to set proper wikia.com or fandom.com domain. No need to move into app code.

File: src/ad-engine/utils/geo.js (hunk 5 above)
Comment (dmnsobczak) [reply src/ad-engine/utils/geo.js]: Yes, you right, I missed that part. Unfortunately we have rewrite cookie every pageview to make it up-to-date with session cookie.

File: src/ad-engine/utils/geo.js [hunk 6]
```diff
@@ -1,54 +1,131 @@
...
+
+	if (withCookie) {
+		synchronizeCookie();
+	}
+}
+
+function readSession() {
+	let sid = sessionId || context.get('options.session.id');
+
+	if (!sid) {
```
Comment (fraszczakszymon) [inline src/ad-engine/utils/geo.js]: What if user is inactive for an hour on mobile-wiki?

File: src/ad-engine/utils/geo.js [hunk 7]
```diff
@@ -1,54 +1,131 @@
...
+		withCookie
 	};
+
+	if (withCookie) {
+		synchronizeCookie();
+	}
+}
+
+function readSession() {
+	let sid = sessionId || context.get('options.session.id');
```
Comment (fraszczakszymon) [inline src/ad-engine/utils/geo.js]: Can we have single getter method which is the only one source of truth? We should re-check value on page change event

Comment (dmnsobczak) [reply src/ad-engine/utils/geo.js]: OK, done.

**Pull request merged**